include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -544,6 +544,30 @@ def NarrowMuxRHS : Pat <
   (MoveNameHint $old, (CatPrimOp (MuxPrimOp $cond, $a, $c), $b)),
   [(EqualTypes $lhs, $rhs), (KnownWidth $lhs)]>;
 
+// mux(eq(a, b), a, b) -> b
+def MuxEQOperands : Pat<
+  (MuxPrimOp:$old (EQPrimOp $a, $b), $a, $b),
+  (replaceWithValue $b),
+  [(EqualTypes $a, $b), (KnownWidth $a)]>;
+
+// mux(eq(b, a), a, b) -> b
+def MuxEQOperandsSwapped : Pat<
+  (MuxPrimOp:$old (EQPrimOp $b, $a), $a, $b),
+  (replaceWithValue $b),
+  [(EqualTypes $a, $b), (KnownWidth $a)]>;
+
+// mux(neq(a, b), a, b) -> a
+def MuxNEQOperands : Pat<
+  (MuxPrimOp:$old (NEQPrimOp $a, $b), $a, $b),
+  (replaceWithValue $a),
+  [(EqualTypes $a, $b), (KnownWidth $a)]>;
+
+// mux(neq(b, a), a, b) -> a
+def MuxNEQOperandsSwapped : Pat<
+  (MuxPrimOp:$old (NEQPrimOp $b, $a), $a, $b),
+  (replaceWithValue $a),
+  [(EqualTypes $a, $b), (KnownWidth $a)]>;
```

Comment:
As an alternative to this, does it make sense to canonicalize `mux(neq(a, b), x, y) -> mux(eq(a, b), y, x)`? This would then enable re-use of the first two canonicalizers.

Reply:
That works too. I feel like just replacing the op with the value is more straightforward in this case. But I can change it if you prefer.

Reply:
Go for the `neq -> eq` canonicalization. This has the benefit of both defining a canonical form for mux and producing slightly more readable output. Subjectively:

```verilog
assign e = a != b ? c : d;
```

is worse than:

```verilog
assign e = a == b ? d : c;
```

Reply:
Updated to use `neq -> eq`. I also separated the tests to their own firrtl module.